kratos/utilities/smp_storage.h

```diff
@@ -0,0 +1,95 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#if !defined(KRATOS_SMP_STORAGE_H_INCLUDED )
+#define  KRATOS_SMP_STORAGE_H_INCLUDED
+
+// System includes
+#ifdef KRATOS_SMP_OPENMP
+#include <vector>
+#include "utilities/openmp_utils.h"
+#endif
+
+// Project includes
+#include "utilities/parallel_utilities.h"
+
+namespace Kratos
+{
+///@addtogroup KratosCore
+///@{
+
+///@name Kratos Classes
+///@{
+
+template<class TStorageType>
+class SMPStorage
+{
+public:
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    SMPStorage(){
+        #ifdef KRATOS_SMP_OPENMP
+            mThreadLocalStorageContainer.resize(ParallelUtilities::GetNumThreads());
+        #endif
+    }
+
+    /// Destructor.
+    ~SMPStorage() = default;
+
+    /// Deleted copy constructor
+    SMPStorage(SMPStorage const& rOther) = delete;
+
+    /// Deleted assignment operator
+    SMPStorage& operator=(SMPStorage const& rOther) = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    inline TStorageType& GetThreadLocalStorage()
+    {
+        #ifdef KRATOS_SMP_OPENMP
+            return mThreadLocalStorageContainer[OpenMPUtils::ThisThread()];
+        #elif defined(KRATOS_SMP_CXX11)
+            thread_local TStorageType tls;
```

Comment:
This is really weird. I know ```thread_local``` modifies the lifetime of ```tls``` but it still looks like you're returning a dangling reference. How about making ```tls``` a proper member variable?
```cpp
#ifdef KRATOS_SMP_OPENMP
    std::vector<TStorageType> mThreadLocalStorageContainer;
#elif defined(KRATOS_SMP_CXX11)
    thread_local static TStorageType tls;
#else
    TStorageType mThreadLocalStorageContainer;
#endif
```
(+ obviously changing the name of ```tls``` to conform to Kratos' style guide)

Reply:
I think thread_local implies static but not sure

Also I wanted to do the same a while ago, but I was quite sure that thread_local does not work for that, but unfortunately I dont remember the reason, so I might be wrong.

@RiccardoRossi also did some tests with it, maybe he remembers?

Reply:
As far as i understand, `thread_local` creates static variables when it is called from a thread. So where this is declared matters because, as in the given example, `mSMPStorage` is created when the `Scheme` is created in a serial execution (no threading is allowed for creation of `Scheme`. In the case of @matekelemen suggestion, it will create one `thread_local` variable in the master thread, and it will live untill the termination of the master thread (i.e. termination of the application). Then when this is requested from a parallelized block, all the threads will retireve master thread's thread local storage creating a race condition.

In the case of my suggestion, the `thread_local` variable will be created everytime a parallellized block is executed and also called in parallel creating thread specific local storage for each thread which is the intended behaviour.

I will try to run this through custom made C11 threading for Kratos for testing to check whether the CXX11 threading will work with this. May be if @RiccardoRossi has done something in this regard, that would be very helpfull to know. :)

Reply:
Ok I have done the testing. Followings are the observations:

1. defining a local member variable as `thread_local` for a class is not allowed. (https://en.cppreference.com/w/cpp/language/storage_duration)
2. The proposed methodology works with C++17. Following is the code block which i used to test.

```c++
#include <vector>
#include <algorithm>
#include <numeric>
#include <execution>
#include <iostream>
#include <mutex>
#include <iostream>

int main()
{
    class Scheme
    {
    public:
        void CalculateLocalSystem(int& LHS, int ProcessInfo) {
            // following line gives race condition as expected.
            // auto& tls = mTLS;

            // following works as a thread_local storage
            thread_local TLS tls;

            tls.b = ProcessInfo * 3;
            LHS = tls.b;
        }
    private:
        struct TLS
        {
            int b;
        };

        TLS mTLS;
    };

    Scheme scheme;

    int v = 0;
    std::mutex m;

    int n = 1e+4;
    std::vector<int> data(n, 0);
    std::iota(data.begin(), data.end(), 0);

    std::for_each(std::execution::par, data.begin(), data.end(), [&](int& ProcessInfo) {
        int lhs;
        scheme.CalculateLocalSystem(lhs, ProcessInfo);
        m.lock();
        v += lhs;
        m.unlock();
    });

    std::cout << "v = " << v << std::endl;
}
```

compiled with `g++  --std=c++17 cxx_11_par.cpp -ltbb`

Any comments would be nice :)

Reply:
@pooyan-dadvand not exactly, `thread_local` will make a thread local static variable for the context. So, if you run the code snippet, it will create TLS objects for each thread, not for each call. You will see it with the cout clearly.

Reply:
`thread_local` implies `static` as well

Reply:
just to quickly sneak in into this discussion,

in my opinion the way we use thread local stuff in scheme should be avoided. I wonder if we should modify the scheme so that instead of doing in builder and solver

~~~
loop on elements:
      scheme(elem.LeftHandSide)
      Assemble
~~~

we call the builder and solver FROM the scheme so that the loop on elements is in the scheme instead in the builder and solver.

that way each scheme could implement more optimally the tls storage it needs by just rewriting the element loop